juniper_codegen/src/util/tracing.rs

```diff
@@ -0,0 +1,352 @@
+    fn raw_name(&self) -> &syn::Ident;
+}
+
+fn is_traced(ty: &impl TracedType, field: &impl TracedField) -> bool {
+    let traced = ty
+        .tracing_rule()
+        .map_or_else(|| true, |rule| rule.is_traced(field));
+
+    let no_trace = field.tracing_attr().map(|t| t.no_trace).unwrap_or(false);
+
+    traced && !no_trace
+}
+
+pub fn instrument() -> TokenStream {
+    quote!(
+        use ::juniper::InstrumentInternal as _;
+    )
+}
+
+// Returns code that constructs `span` required for tracing
+pub fn span_tokens(ty: &impl TracedType, field: &impl TracedField) -> TokenStream {
+    if !is_traced(ty, field) {
+        return quote!();
+    }
+
+    let name = field.name();
+    let span_name = format!("{}.{}", ty.name(), name);
+    let span_name = syn::LitStr::new(&span_name, name.span());
+
+    let args = field.args().into_iter().filter_map(|arg| {
+        let name = arg.name();
+        let raw_name = arg.raw_name();
+        let arg_name = syn::LitStr::new(name, arg.ty().span());
+
+        field
+            .tracing_attr()
+            .map(|t| t.skip.get(&raw_name.to_string()))
+            .flatten()
+            .is_none()
+            .then(|| {
+                quote!(
+                    #arg_name = ::juniper::tracing::field::debug(&#raw_name)
+                )
+            })
+    });
+
+    let args: Vec<_> = if let Some(tracing) = field.tracing_attr() {
+        let additional_fields = tracing.fields.iter().map(|f| {
+            let name = &f.left;
+            let right = &f.right;
+            quote!(#name = ::juniper::tracing::field::debug(#right))
+        });
+
+        args.chain(additional_fields).collect()
+    } else {
+        args.collect()
+    };
+
+    let level = field
+        .tracing_attr()
+        .map(|t| t.level.as_ref())
+        .flatten()
+        .map(|l| match l.value().as_str() {
+            "trace" => quote!(TRACE),
+            "debug" => quote!(DEBUG),
+            "info" => quote!(INFO),
+            "warn" => quote!(WARN),
+            "error" => quote!(ERROR),
+            l => abort!(syn::Error::new(
+                l.span(),
+                format!(
+                    "Unsupported tracing level: {}, \
+                     supported values: trace, debug, info, warn, error",
+                    l,
+                ),
+            )),
+        })
+        .unwrap_or_else(|| quote!(INFO));
+
+    let target = field
+        .tracing_attr()
+        .map(|t| t.target.as_ref())
+        .flatten()
+        .map_or_else(|| quote!(), |t| quote!(target: #t,));
+
+    quote!(
+        let _tracing_span = ::juniper::tracing::span!(
+            #target ::juniper::tracing::Level::#level, #span_name, #(#args, )*
+        );
+    )
+}
+
+// Returns code to start tracing of async future
+pub fn async_tokens(ty: &impl TracedType, field: &impl TracedField) -> TokenStream {
+    if !is_traced(ty, field) {
+        return quote!();
+    }
+    quote!(.__instrument(_tracing_span))
```

Comment:
@ArsileLuci yup, as expected. Use universal trait method syntax call, and you won't need that custom trait. Like that:
```rust
let f = ::juniper::tracing::Instrument(f, _tracing_span);
```

Reply:
@tyranron This becomes rather complicated once we introduce streams, to trace subscriptions, and in my opinion it' much easier to slap extension trait where it needs and keep it as simple as possible.